### PR TITLE
Closes a number of issues identified by @alebastrov

### PR DESCRIPTION
**Pull request by adamretter**

### CODE REVIEW

File: extensions/webdav/src/org/exist/webdav/ExistCollection.java [hunk 1]
```diff
@@ -368,10 +368,15 @@ public XmldbURI createFile(String newName, InputStream is, Long length, String c
 
                 // Stream into database
                 VirtualTempFileInputSource vtfis = new VirtualTempFileInputSource(vtf);
-                IndexInfo info = collection.validateXMLResource(txn, broker, newNameUri, vtfis);
-                DocumentImpl doc = info.getDocument();
-                doc.getMetadata().setMimeType(mime.getName());
-                collection.store(txn, broker, info, vtfis, false);
+                try {
```
Comment (dizzzz) [inline extensions/webdav/src/org/exist/webdav/ExistCollection.java]: probably `try(VirtualTempFileInputSource vtfis = new VirtualTempFileInputSource(vtf);)` would have been better here.

File: src/org/exist/http/RESTServer.java [hunk 2]
```diff
@@ -1069,12 +1069,16 @@ public void doPut(final DBBroker broker, final XmldbURI path,
             }
 
             if (mime.isXMLType()) {
-                final InputSource vtfis = new VirtualTempFileInputSource(vtempFile, charset);
-
-                final IndexInfo info = collection.validateXMLResource(transaction, broker, docUri, vtfis);
-                info.getDocument().getMetadata().setMimeType(contentType);
-                collection.store(transaction, broker, info, vtfis, false);
-                response.setStatus(HttpServletResponse.SC_CREATED);
+                final VirtualTempFileInputSource vtfis = new VirtualTempFileInputSource(vtempFile, charset);
```
Comment (dizzzz) [inline src/org/exist/http/RESTServer.java]: Should have neen `try(resource)` too?

File: src/org/exist/security/AbstractAccount.java [hunk 3]
```diff
@@ -246,20 +246,18 @@ public final String toString() {
...
-        } else {
-            other = null;
-        }
-
-        if (other != null){
-            return (getRealm() == other.getRealm() && name.equals(other.name));
-        } //id == other.id;
-
-        return false;
+        return Optional
```
Comment (dizzzz) [inline src/org/exist/security/AbstractAccount.java]: brilliant !

File: src/org/exist/security/AbstractSubject.java [hunk 4]
```diff
@@ -176,7 +177,11 @@ public boolean isEnabled() {
 
     @Override
     public boolean equals(final Object obj) {
-        return account.equals(obj);
+       return Optional
```
Comment (dizzzz) [inline src/org/exist/security/AbstractSubject.java]: is this semantically actually the same?

File: src/org/exist/storage/BrokerPool.java [hunk 5]
```diff
@@ -2039,6 +2046,51 @@ public void shutdown(final boolean killed) {
         }
     }
 
+    private void clearThreadLocals() {
```
Comment (dizzzz) [inline src/org/exist/storage/BrokerPool.java]: I trust you on this one! Thnx for the code comments!

File: src/org/exist/util/VirtualTempFileInputSource.java [hunk 6]
```diff
@@ -37,104 +43,76 @@
...
-				{bs = vtempFile.getByteStream();}
-			else if(file!=null)
-				{bs = new BufferedInputStream(new FileInputStream(file));}
-		} catch(final IOException ioe) {
-			// DoNothing(R)
-		}
-		
-		return bs;
+		return file
+				.flatMap(f -> f.fold(this::newInputStream, this::vtfByteStream))
```
Comment (dizzzz) [inline src/org/exist/util/VirtualTempFileInputSource.java]: Did you remove here the BufferedInputStream?

File: src/org/exist/util/function/Either.java [hunk 7]
```diff
@@ -52,18 +53,43 @@ public final boolean isRight() {
     public final RightProjection<L, R> right() {
         return new RightProjection(this);
     }
-    
+
+    /**
+     * Catamorphism. Run the first given function if left,
```
Comment (dizzzz) [inline src/org/exist/util/function/Either.java]: interesting, though I do not yet fully understand the purpose. Looks like a neat clean construct to me....

File: src/org/exist/security/AbstractSubject.java (hunk 4 above)
Comment (adamretter) [inline src/org/exist/security/AbstractSubject.java]: Yup

File: src/org/exist/util/function/SupplierE.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+/**
+ * Similar to {@link java.util.function.Supplier} but
+ * permits a statically know Exception to be thrown
```
Comment (dizzzz) [inline src/org/exist/util/function/SupplierE.java]: Exception or.... Throwable?

File: src/org/exist/util/VirtualTempFileInputSource.java (hunk 6 above)
Comment (adamretter) [inline src/org/exist/util/VirtualTempFileInputSource.java]: Nope, it just moved into the newInputStream() method.

File: src/org/exist/xmldb/RemoteResourceSet.java [hunk 9]
```diff
@@ -235,7 +235,8 @@ public void removeResource(final long pos) throws XMLDBException {
     protected void finalize() throws Throwable {
         try {
             clear();
-        } finally {
+        }
```
Comment (dizzzz) [inline src/org/exist/xmldb/RemoteResourceSet.java]: what formatting rules do you use? i think in the code base `} finally {' is done on most locations....

File: src/org/exist/util/VirtualTempFileInputSource.java (hunk 6 above)
Comment (dizzzz) [inline src/org/exist/util/VirtualTempFileInputSource.java]: right you are :-)

File: extensions/webdav/src/org/exist/webdav/ExistCollection.java (hunk 1 above)
Comment (adamretter) [inline extensions/webdav/src/org/exist/webdav/ExistCollection.java]: Indeed. Sorry that was cherry-picked from @alebastrov. I will fix and rebase.

File: src/org/exist/http/RESTServer.java (hunk 2 above)
Comment (adamretter) [inline src/org/exist/http/RESTServer.java]: Again the same as above:
> Indeed. Sorry that was cherry-picked from @alebastrov. I will fix and rebase.

File: src/org/exist/security/AbstractSubject.java (hunk 4 above)
Comment (dizzzz) [inline src/org/exist/security/AbstractSubject.java]: ok thnx !

File: extensions/webdav/src/org/exist/webdav/ExistCollection.java (hunk 1 above)
Comment (dizzzz) [inline extensions/webdav/src/org/exist/webdav/ExistCollection.java]: I understand :-) I found one other instance , marked below.

File: src/org/exist/http/RESTServer.java (hunk 2 above)
Comment (dizzzz) [inline src/org/exist/http/RESTServer.java]: Ok :-)

File: src/org/exist/xmldb/RemoteResourceSet.java (hunk 9 above)
Comment (adamretter) [inline src/org/exist/xmldb/RemoteResourceSet.java]: Again the same as above:
> Indeed. Sorry that was cherry-picked from @alebastrov. I will fix and rebase.

**Comment by dizzzz** [comment]
Thnx @adamretter for the changes, they look good to me, again I learn form your code. Just a few questions/remarks, which you are already picking up, super!

**Comment by adamretter** [comment]
@dizzzz Just addressed your issues and rebased. Should be good to go now...

**Pull request merged**